File: src/BinarySearchTree.java
// Implements the Binary Search Tree data structure
public class BinarySearchTree {

    // Inner class to represent a node in the BST
    private static class Node {
        int value;
        Node left;
        Node right;

        // Constructor to create a new node
        Node(int value) {
            this.value = value;
            left = null;
            right = null;
        }
    }

    private Node root;

    public BinarySearchTree() {
        root = null;
    }

    // --- Task A1: BST Core Operations ---

    // Inserts a value into the BST
    public void insert(int value) {
        root = insertRec(root, value);
    }

    // A recursive function to insert a new value in BST
    private Node insertRec(Node current, int value) {
        // If the tree is empty, return a new node
        if (current == null) {
            return new Node(value);
        }

        // Otherwise, recur down the tree
        if (value < current.value) {
            current.left = insertRec(current.left, value);
        } else if (value > current.value) {
            current.right = insertRec(current.right, value);
        }

        // return the (unchanged) node pointer
        return current;
    }

    // Searches for a value in the BST
    public boolean search(int value) {
        return searchRec(root, value);
    }

    // A recursive function to search for a value in BST
    private boolean searchRec(Node current, int value) {
        // Base case: root is null or value is present at root
        if (current == null) {
            return false;
        }
        if (current.value == value) {
            return true;
        }

        // Value is greater than root's value
        if (value < current.value) {
            return searchRec(current.left, value);
        }

        // Value is less than root's value
        return searchRec(current.right, value);
    }

    // Deletes a value from the BST
    public void delete(int value) {
        root = deleteRec(root, value);
    }

    // A recursive function to delete a value in BST
    private Node deleteRec(Node root, int value) {
        // Base Case: If the tree is empty
        if (root == null) {
            return root;
        }

        // Otherwise, recur down the tree
        if (value < root.value) {
            root.left = deleteRec(root.left, value);
        } else if (value > root.value) {
            root.right = deleteRec(root.right, value);
        } else {
            // node with only one child or no child
            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            }

            // node with two children: Get the inorder successor (smallest in the right subtree)
            root.value = minValue(root.right);

            // Delete the inorder successor
            root.right = deleteRec(root.right, root.value);
        }

        return root;
    }

    private int minValue(Node root) {
        int minv = root.value;
        while (root.left != null) {
            minv = root.left.value;
            root = root.left;
        }
        return minv;
    }

    // Finds the minimum value in the BST
    public int findMin() {
        if (root == null) {
            throw new IllegalStateException("Tree is empty");
        }
        Node current = root;
        while (current.left != null) {
            current = current.left;
        }
        return current.value;
    }

    // Finds the maximum value in the BST
    public int findMax() {
        if (root == null) {
            throw new IllegalStateException("Tree is empty");
        }
        Node current = root;
        while (current.right != null) {
            current = current.right;
        }
        return current.value;
    }

    // --- Task A2: Tree Traversal ---

    // Displays the inorder traversal of the BST
    public void printInorder() {
        System.out.print("Inorder Traversal: ");
        printInorderRec(root);
        System.out.println();
    }

    private void printInorderRec(Node node) {
        if (node != null) {
            printInorderRec(node.left);
            System.out.print(node.value + " ");
            printInorderRec(node.right);
        }
    }

    // Displays the preorder traversal of the BST
    public void printPreorder() {
        System.out.print("Preorder Traversal: ");
        printPreorderRec(root);
        System.out.println();
    }

    private void printPreorderRec(Node node) {
        if (node != null) {
            System.out.print(node.value + " ");
            printPreorderRec(node.left);
            printPreorderRec(node.right);
        }
    }

    // Displays the postorder traversal of the BST
    public void printPostorder() {
        System.out.print("Postorder Traversal: ");
        printPostorderRec(root);
        System.out.println();
    }

    private void printPostorderRec(Node node) {
        if (node != null) {
            printPostorderRec(node.left);
            printPostorderRec(node.right);
            System.out.print(node.value + " ");
        }
    }

    // --- Task A3: Tree Properties ---

    // Calculates the height of the tree
    public int getHeight() {
        return getHeightRec(root);
    }

    private int getHeightRec(Node node) {
        if (node == null) {
            return -1; // Height of empty tree is -1 (or 0 depending on convention, using -1 for edge-count definition)
        }
        int leftHeight = getHeightRec(node.left);
        int rightHeight = getHeightRec(node.right);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    // Counts the total number of nodes in the tree
    public int getTotalNodes() {
        return getTotalNodesRec(root);
    }

    private int getTotalNodesRec(Node node) {
        if (node == null) {
            return 0;
        }
        return 1 + getTotalNodesRec(node.left) + getTotalNodesRec(node.right);
    }

    // Counts the number of leaf nodes in the tree
    public int getLeafNodes() {
        return getLeafNodesRec(root);
    }

    private int getLeafNodesRec(Node node) {
        if (node == null) {
            return 0;
        }
        if (node.left == null && node.right == null) {
            return 1;
        }
        return getLeafNodesRec(node.left) + getLeafNodesRec(node.right);
    }
}

File: python/BinarySearchTree.py
class Node:
    def __init__(self, key):
        self.val = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    # --- Task A1: BST Core Operations ---

    def insert(self, key):
        self.root = self._insert_recursive(self.root, key)

    def _insert_recursive(self, root, key):
        if root is None:
            return Node(key)
        if key < root.val:
            root.left = self._insert_recursive(root.left, key)
        elif key > root.val:
            root.right = self._insert_recursive(root.right, key)
        return root

    def delete(self, key):
        self.root = self._delete_recursive(self.root, key)

    def _delete_recursive(self, root, key):
        if root is None:
            return root
        if key < root.val:
            root.left = self._delete_recursive(root.left, key)
        elif key > root.val:
            root.right = self._delete_recursive(root.right, key)
        else:
            # Node with only one child or no child
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            
            # Node with two children: Get the inorder successor (smallest in the right subtree)
            root.val = self._min_value(root.right)
            # Delete the inorder successor
            root.right = self._delete_recursive(root.right, root.val)
        return root

    def search(self, key):
        return self._search_recursive(self.root, key)

    def _search_recursive(self, root, key):
        if root is None:
            return False
        if root.val == key:
            return True
        if key < root.val:
            return self._search_recursive(root.left, key)
        return self._search_recursive(root.right, key)

    def _min_value(self, root):
        min_v = root.val
        while root.left:
            min_v = root.left.val
            root = root.left
        return min_v

    def find_min(self):
        if self.root is None:
            return None
        return self._min_value(self.root)

    def find_max(self):
        if self.root is None:
            return None
        current = self.root
        while current.right:
            current = current.right
        return current.val

    # --- Task A2: Tree Traversal ---

    def inorder_traversal(self):
        result = []
        self._inorder_recursive(self.root, result)
        return result

    def _inorder_recursive(self, root, result):
        if root:
            self._inorder_recursive(root.left, result)
            result.append(root.val)
            self._inorder_recursive(root.right, result)

    def preorder_traversal(self):
        result = []
        self._preorder_recursive(self.root, result)
        return result

    def _preorder_recursive(self, root, result):
        if root:
            result.append(root.val)
            self._preorder_recursive(root.left, result)
            self._preorder_recursive(root.right, result)

    def postorder_traversal(self):
        result = []
        self._postorder_recursive(self.root, result)
        return result

    def _postorder_recursive(self, root, result):
        if root:
            self._postorder_recursive(root.left, result)
            self._postorder_recursive(root.right, result)
            result.append(root.val)

    # --- Task A3: Tree Properties ---

    def height(self):
        return self._height_recursive(self.root)

    def _height_recursive(self, root):
        if root is None:
            return -1 # Matching Java implementation convention
        left_height = self._height_recursive(root.left)
        right_height = self._height_recursive(root.right)
        return max(left_height, right_height) + 1

    def node_count(self):
        return self._node_count_recursive(self.root)

    def _node_count_recursive(self, root):
        if root is None:
            return 0
        return 1 + self._node_count_recursive(root.left) + self._node_count_recursive(root.right)

    def leaf_count(self):
        return self._leaf_count_recursive(self.root)

    def _leaf_count_recursive(self, root):
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return 1
        return self._leaf_count_recursive(root.left) + self._leaf_count_recursive(root.right)
